### PR TITLE
[jenkins] fix: add support to publish artifacts to Nexus public and private repo base on GitHub repo visibility

### PR DESCRIPTION
**Pull request by Wayonb**
problem: currently, Jenkins publishes artifacts to a Nexus repo without knowing its visibility
solution: Jenkins will now publish private GitHub repos to a private Nexus repo

### CODE REVIEW

File: jenkins/shared-library/vars/configureArtifactRepository.groovy [hunk 1]
```diff
@@ -1,25 +1,40 @@
...
 
-	String url = resolveRepositoryUrl(ownerName, resolveRepositoryName(environment))
+	final String url = resolveRepositoryUrl(ownerName, resolveRepositoryName(environment, publicGitHubRepo))
 
 	if (null != url) {
 		configure(environment, ownerName, url)
+
+		if ('javascript' == environment && fileExists('package-lock.json')) {
+			// remove package-lock.json file since the hashes will not match private repository
+			sh('rm -f package-lock.json')
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/configureArtifactRepository.groovy:12]: why this ?

File: jenkins/shared-library/vars/helper.groovy [hunk 2]
```diff
@@ -107,3 +107,13 @@ String resolveWorkspacePath(String os) {
 String resolveGitHubCredentialsId() {
 	return scm.userRemoteConfigs[0].credentialsId
 }
+
+void withTempDir(Closure body) {
+	dir( pwd(tmp: true) ) {
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/helper.groovy:112]: extra spaces in dir parens?

File: jenkins/shared-library/vars/publish.groovy [hunk 3]
```diff
@@ -94,14 +98,14 @@ void pythonPublisher(Map config, String phase) {
 		return
 	}
 
-	final String ownerName = helper.resolveOrganizationName()
-	final String repositoryName = helper.resolveRepositoryName()
-	if (isAlphaRelease(phase) || !githubHelper.isGitHubRepositoryPublic(ownerName, repositoryName)) {
+	if (isAlphaRelease(phase) || !config.publicGitHubRepo) {
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/publish.groovy:101]: publicGitHubRepo => isPublicGitHubRepo (assuming boolean) ?

File: jenkins/shared-library/vars/configureArtifactRepository.groovy [hunk 4]
```diff
@@ -1,25 +1,40 @@
...
+	}
+
+	return publicGitHubRepo ? repositoryName : "${repositoryName}-private"
 }
 
 String readNpmPackageScopeName() {
 	Object packageJson = readJSON file: 'package.json'
+	if (null == packageJson.name) {
+		return null
+	}
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/configureArtifactRepository.groovy:36]: can this happen in normal operation?

File: jenkins/shared-library/vars/githubHelper.groovy [hunk 5]
```diff
@@ -2,24 +2,42 @@ import groovy.json.JsonOutput
...
+}
+
 String executeGithubApiRequest(String command) {
-	final String response = runScript(command, true)
-	if (response.contains('message')) {
-		println "GitHub API request failed: ${response}"
-		throw new IllegalArgumentException(response)
+	String results = ''
+	helper.withTempDir {
+		String file = "./${System.currentTimeMillis()}"
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/githubHelper.groovy:31]: why are you spooling to temp file ?

File: jenkins/shared-library/vars/publish.groovy [hunk 6]
```diff
@@ -64,10 +68,10 @@ void npmPublisher(Map config, String phase) {
 		npmPublishCommand.append(' --tag alpha')
 	}
 
-	final String ownerName = helper.resolveOrganizationName()
-	final String repositoryName = helper.resolveRepositoryName()
-	if (isAlphaRelease(phase) || !githubHelper.isGitHubRepositoryPublic(ownerName, repositoryName)) {
-		final String publishUrl = configureArtifactRepository.resolveRepositoryUrl(ownerName, 'npm-hosted')
+	if (isAlphaRelease(phase) || !config.publicGitHubRepo) {
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/publish.groovy:71]: consider refactoring this condition (used three times)

File: jenkins/shared-library/vars/configureArtifactRepository.groovy (hunk 1 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/configureArtifactRepository.groovy:12]: When trying to update from internal packages such as symbol-sdk, it gets a hash error due to hash does not match what’s in the package-lock.json file.

File: jenkins/shared-library/vars/githubHelper.groovy (hunk 5 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/githubHelper.groovy:31]: The output from the GitHub API is saved to a temp file, while the HTTP response code is displayed on the screen.
Separate these, so I don't need to parse the output from the screen.

File: jenkins/shared-library/vars/configureArtifactRepository.groovy (hunk 1 above)
Comment (Jaguar0625) [reply jenkins/shared-library/vars/configureArtifactRepository.groovy:12]: you mean projects depending on private artifacts, i.e. using symbol-sdk?

**Review by Jaguar0625 — APPROVED** [review]
👍

File: jenkins/shared-library/vars/configureArtifactRepository.groovy (hunk 1 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/configureArtifactRepository.groovy:12]: yea, so the package-lock.json file resolved the package info to npmjs.com with their hash.  I guess this is different from hash in Nexus(due to Metadata).  Removing the package-lock.json will fix the hash error.

**Pull request merged**